File: src/main/java/com/oyosite/ticon/specutils/block/NoxwoodDeco.java
package com.oyosite.ticon.specutils.block;

import com.oyosite.ticon.specutils.*;
import de.dafuqs.spectrum.blocks.decoration.*;
import de.dafuqs.spectrum.registries.*;
import net.minecraft.core.*;
import net.minecraft.core.registries.*;
import net.minecraft.resources.*;
import net.minecraft.util.*;
import net.minecraft.world.item.*;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.state.*;
import net.minecraft.world.level.material.*;

import java.util.*;

public class NoxwoodDeco {

	public enum NoxwoodType {
		CHESTNUT, SLATE, EBONY, IVORY;

		public static final List<Tuple<NoxwoodType, NoxwoodType>> COMBINATIONS = getCombinations();

		@Override
		public String toString() {
			return name().toLowerCase();
		}

		public static List<Tuple<NoxwoodType, NoxwoodType>> getCombinations() {
			List<Tuple<NoxwoodType, NoxwoodType>> results = new ArrayList<>();
			for (NoxwoodType type1 : values()) {
				for (NoxwoodType type2 : values()) {
					if (type1 != type2) {
						results.add(new Tuple<>(type1, type2));
					}
				}
			}
			return results;
		}
	}

	public static final Map<ResourceLocation, Block> LAMPS = new LinkedHashMap<>();
	public static final Map<ResourceLocation, Block> LANTERNS = new LinkedHashMap<>();
	public static final Map<ResourceLocation, Block> LIGHTS = new LinkedHashMap<>();
	public static final Map<ResourceLocation, Block> ALL_DECO = new LinkedHashMap<>();

	static {
		for (Tuple<NoxwoodType, NoxwoodType> pair : NoxwoodType.COMBINATIONS) {
			/*ResourceLocation lampId = idFor(pair, "lamp");
			Block lampBlock = noxwoodLampBlock();
			LAMPS.put(lampId, lampBlock);
			ALL_DECO.put(lampId, lampBlock);*/

			ResourceLocation lanternId = idFor(pair, "lantern");
			Block lanternBlock = noxwoodLanternBlock();
			LANTERNS.put(lanternId, lanternBlock);
			ALL_DECO.put(lanternId, lanternBlock);

			/*ResourceLocation lightId = idFor(pair, "light");
			Block lightBlock = noxwoodLightBlock();
			LIGHTS.put(lightId, lightBlock);
			ALL_DECO.put(lightId, lightBlock);*/
		}
	}

	public static void registerAll() {
		for (Map.Entry<ResourceLocation, Block> entry : ALL_DECO.entrySet()) {
			ResourceLocation id = entry.getKey();
			Block block = entry.getValue();
			Registry.register(BuiltInRegistries.BLOCK, id, block);
			Registry.register(BuiltInRegistries.ITEM, id, new BlockItem(block, SpectrumItems.IS.of()));
		}
	}

	public static ResourceLocation idFor(Tuple<NoxwoodType, NoxwoodType> pair, String blockName) {
		return SpectralUtilities.id(
				pair.getA().name().toLowerCase() + "_" +
						pair.getB().name().toLowerCase() + "_noxwood_" + blockName
		);
	}

	public static Block noxwoodLampBlock() {
		return new RedstoneLampBlock(
				SpectrumBlocks.noxcap(MapColor.CRIMSON_NYLIUM)
						.lightLevel(SpectrumBlocks.LANTERN_LIGHT_PROVIDER)
		);
	}

	public static Block noxwoodLanternBlock() {
		return new FlexLanternBlock(
				BlockBehaviour.Properties.ofFullCopy(Blocks.LANTERN)
						.lightLevel(value -> 13)
						.pushReaction(PushReaction.DESTROY)
		);
	}

	public static Block noxwoodLightBlock() {
		return new RotatedPillarBlock(
				SpectrumBlocks.noxcap(MapColor.CRIMSON_NYLIUM).lightLevel(state -> 15)
		);
	}

}

File: src/main/kotlin/com/oyosite/ticon/specutils/block/NoxwoodDeco.kt
package com.oyosite.ticon.specutils.block

import com.oyosite.ticon.specutils.SpectralUtilities.Companion.id
import de.dafuqs.spectrum.blocks.decoration.FlexLanternBlock
import de.dafuqs.spectrum.registries.SpectrumBlocks
import de.dafuqs.spectrum.registries.SpectrumItems
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.Tuple
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.RedstoneLampBlock
import net.minecraft.world.level.block.RotatedPillarBlock
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.material.MapColor
import net.minecraft.world.level.material.PushReaction
import java.util.*
import java.util.function.ToIntFunction

object NoxwoodDeco {
    @JvmField
	val LAMPS: MutableMap<ResourceLocation?, Block?> = LinkedHashMap<ResourceLocation?, Block?>()
    val LANTERNS: MutableMap<ResourceLocation?, Block?> = LinkedHashMap<ResourceLocation?, Block?>()
    val LIGHTS: MutableMap<ResourceLocation?, Block?> = LinkedHashMap<ResourceLocation?, Block?>()
    val ALL_DECO: MutableMap<ResourceLocation?, Block?> = LinkedHashMap<ResourceLocation?, Block?>()

    init {
        for (pair in NoxwoodType.Companion.COMBINATIONS) {
            /*ResourceLocation lampId = idFor(pair, "lamp");
			Block lampBlock = noxwoodLampBlock();
			LAMPS.put(lampId, lampBlock);
			ALL_DECO.put(lampId, lampBlock);*/

            val lanternId = idFor(pair, "lantern")
            val lanternBlock = noxwoodLanternBlock()
            LANTERNS.put(lanternId, lanternBlock)
            ALL_DECO.put(lanternId, lanternBlock)

            /*ResourceLocation lightId = idFor(pair, "light");
			Block lightBlock = noxwoodLightBlock();
			LIGHTS.put(lightId, lightBlock);
			ALL_DECO.put(lightId, lightBlock);*/
        }
    }

    @JvmStatic
	fun registerAll() {
        for (entry in ALL_DECO.entries) {
            val id: ResourceLocation = entry.key!!
            val block: Block = entry.value!!
            Registry.register<Block?, Block?>(BuiltInRegistries.BLOCK, id, block)
            Registry.register<Item?, BlockItem?>(BuiltInRegistries.ITEM, id, BlockItem(block, SpectrumItems.IS.of()))
        }
    }

    fun idFor(pair: Tuple<NoxwoodType?, NoxwoodType?>, blockName: String?): ResourceLocation {
        return id(
            pair.getA()!!.name.lowercase(Locale.getDefault()) + "_" +
                    pair.getB()!!.name.lowercase(Locale.getDefault()) + "_noxwood_" + blockName
        )
    }

    fun noxwoodLampBlock(): Block {
        return RedstoneLampBlock(
            SpectrumBlocks.noxcap(MapColor.CRIMSON_NYLIUM)
                .lightLevel(SpectrumBlocks.LANTERN_LIGHT_PROVIDER)
        )
    }

    fun noxwoodLanternBlock(): Block {
        return FlexLanternBlock(
            BlockBehaviour.Properties.ofFullCopy(Blocks.LANTERN)
                .lightLevel(ToIntFunction { value: BlockState? -> 13 })
                .pushReaction(PushReaction.DESTROY)
        )
    }

    fun noxwoodLightBlock(): Block {
        return RotatedPillarBlock(
            SpectrumBlocks.noxcap(MapColor.CRIMSON_NYLIUM).lightLevel(ToIntFunction { state: BlockState? -> 15 })
        )
    }

    enum class NoxwoodType {
        CHESTNUT, SLATE, EBONY, IVORY;

        override fun toString(): String {
            return name.lowercase(Locale.getDefault())
        }

        companion object {
            val COMBINATIONS: MutableList<Tuple<NoxwoodType?, NoxwoodType?>> =
                combinations

            val combinations: MutableList<Tuple<NoxwoodType?, NoxwoodType?>>
                get() {
                    val results: MutableList<Tuple<NoxwoodType?, NoxwoodType?>> =
                        ArrayList<Tuple<NoxwoodType?, NoxwoodType?>>()
                    for (type1 in entries) {
                        for (type2 in entries) {
                            if (type1 != type2) {
                                results.add(Tuple<NoxwoodType?, NoxwoodType?>(type1, type2))
                            }
                        }
                    }
                    return results
                }
        }
    }
}